File: day6/go/part_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var grid [][]rune

	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, []rune(line))
	}

	N := len(grid)
	M := len(grid[0])
	var startX, startY int

	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			if grid[i][j] == '^' {
				startY, startX = i, j
				break
			}
		}
	}

	visited := make(map[string]bool)
	visited[fmt.Sprintf("%d,%d", startY, startX)] = true
	queue := [][]int{{startY, startX}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		y, x := curr[0], curr[1]

		for _, dir := range [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}} {
			newY, newX := y+dir[0], x+dir[1]
			if newY >= 0 && newY < N && newX >= 0 && newX < M && grid[newY][newX] != '#' {
				key := fmt.Sprintf("%d,%d", newY, newX)
				if !visited[key] {
					visited[key] = true
					queue = append(queue, []int{newY, newX})
				}
			}
		}
	}

	fmt.Println(len(visited))
}

File: day6/ruby/part_1.rb
grid = File.readlines('input.txt').map(&:chomp)
N, M = grid.size, grid[0].size
start_x = start_y = 0

grid.each_with_index do |row, i|
  row.each_char.with_index do |cell, j|
    if cell == '^'
      start_y, start_x = i, j
      break
    end
  end
end

visited = {[start_y, start_x] => true}
queue = [[start_y, start_x]]
dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

until queue.empty?
  y, x = queue.shift
  dirs.each do |dy, dx|
    new_y, new_x = y + dy, x + dx
    if new_y >= 0 && new_y < N && new_x >= 0 && new_x < M && grid[new_y][new_x] != '#'
      unless visited[[new_y, new_x]]
        visited[[new_y, new_x]] = true
        queue << [new_y, new_x]
      end
    end
  end
end

puts visited.size
